### ISSUE TITLE
Issue with letsencrypt certificate renewal

### ISSUE THREAD

**Issue by jrnoble83** [Ylianst/MeshCentral#4511]
**Describe your issue**

I started to get a warning message that my letsencrypt certificate would expire in "X" days.  I started looking at other posts and trying different fixes.  I first backed up and then removed the /meshctral-data/letsencrypt-certs folder.  When server restarted the folder was recreated but failed to pull down new certs.  I confirmed port forwarding specifically on port 80 and 443 and verified that was working.  I then checked and discovered I had to reset permissions for MC to use ports below 1025.  I fixed that.  I then started MC manually with the -debug cert and got the following:

jnoble@meshcentral-server:~$ node node_modules/meshcentral --debug cert
MeshCentral HTTP redirection server running on port 80.
CERT: LE: Getting certs from local store (Production)
CERT: LE: No certificate files found
MeshCentral v1.0.0, Hybrid (LAN + WAN) mode.
MeshCentral Intel(R) AMT server running on meshcentral.longnoble.org:4433.
MeshCentral HTTPS server running on meshcentral.longnoble.org:443.
SMTP mail server alertserver.mobilcom.com working as expected.
CERT: LE: Got no certificates, asking for one now.
CERT: LE: Generating private key...
CERT: LE: Setting up ACME client...
CERT: LE: Creating certificate request...
CERT: LE: Requesting certificate from Let's Encrypt...
CERT: LE: Failed to obtain certificate: The value "jwk" is invalid for option "format"

I tried to search on the failed message but wasn't able to find anything I could find helpful.

Here is my config.json if it helps anyone point out anything:

{
  "$schema": "http://info.meshcentral.com/downloads/meshcentral-config-schema.json",
  "__comment1__": "This is a simple configuration file, all values and sections that start with underscore (_) are ignored. Edit a section and remove the _ in front of the name. Refer to the user's>
  "__comment2__": "See node_modules/meshcentral/sample-config-advanced.json for a more advanced example.",
  "settings": {
    "MongoDb": "mongodb://127.0.0.1:27017/meshcentral",
    "cert": "meshcentral.longnoble.org",
    "relayDNS": "relaymeshcentral.longnoble.org",
    "_WANonly": true,
    "_LANonly" : true,
    "Port": 443,
    "RedirPort": 80,
    "RelayPort": 453,
    "relayAliasPort": 443,
    "MpsPort": 4433,
    "_ignoreagenthashcheck": true,
    "_AllowLoginToken": true,
    "_AllowFraming": true,
    "_WebRTC": false,
    "_ClickOnce": false,
    "_UserAllowedIP" : "127.0.0.1,::1,192.168.0.100",
    "Compression": true,
    "WsCompression": true,
    "AgentWsCompression": true,
    "DesktopMultiplex": true
  },
  "domains": {
    "": {
      "title": "meshcentral.longnoble.org",
      "_title2": "meshcentral.longnoble.org",
      "minify": true,
      "_newAccounts": true,
      "_userNameIsEmail": true,
      "mstsc": true
    }
  },
  "letsencrypt": {
    "__comment__": "Requires NodeJS 8.x or better, Go to https://letsdebug.net/ first before trying Let's Encrypt.",
    "email": "jrnoble@me.com",
    "names": "meshcentral.longnoble.org",
    "rsaKeySize": 3072,
    "production": true
  },
  "smtp": {
    "host": "alertserver.mobilcom.com",
    "port": 465,
    "from": "meshcentral@longnoble.org",
    "user": "meshcentral@longnoble.org",
    "pass": "****************",
    "__tls__": "When 'tls' is set to true, TLS is used immidiatly when connecting. For SMTP servers that use TLSSTART, set this to 'false' and TLS will still be used.",
    "tls": true,
    "___tlscertcheck__": "When set to false, the TLS certificate of the SMTP server is not checked.",
    "_tlscertcheck": false,
    "__tlsstrict__": "When set to true, TLS cypher setup is more limited, SSLv2 and SSLv3 are not allowed.",
    "_tlsstrict": true
  }
}

**Server Software (please complete the following information):**
 - OS: [Ubuntu 22.04.1 LTS]
 - Virtualization: [None-Whole System is VM on Hyper-V]
 - Network: [LAN/WAN, no reverse proxy, port forwarding directly to system]
 - Version: [Stable 1.0.0]
 - Node: [v12.22.9]

**Comment by si458** [comment]
something is wrong with your `config.json`
remove the following lines
`comment1`
`comment2`
`"_tlscertcheck": "When set to false, the TLS certificate of the SMTP server is not checked."`
`"tlsstrict": "When set to true, TLS cypher setup is more limited, SSLv2 and SSLv3 are not allowed."`
`"tls": "When 'tls' is set to true..."`
lines

these are just comments and not variables, so arent required

**Comment by jrnoble83** [comment]
Thanks for the input, I did clean up the config.json as you recommended.  I am still getting the same error.  After reviewing the sections recommended to be cleaned up that was more with TLS with Email/SMTP server, I don't think that made any changes to the letsencrypt portion of the file.  However I do appreciate the comments to clean up that part.  Thanks

**Comment by mstrhakr** [comment]
Here is how you have Let's Encrypt configured currently.

``` json
"letsencrypt": {
		"comment": "Requires NodeJS 8.x or better, Go to https://letsdebug.net/ first before trying Let's Encrypt.",
		"email": "jrnoble@me.com",
		"names": "meshcentral.longnoble.org",
		"rsaKeySize": 3072,
		"production": true
	},
```
The schema doesn't allow additional properties, so by having the 'comment' and the 'rsaKeySize' properties, you've invalidated the entire 'letsencrypt' object so MeshCentral ignores it.

``` json
"letsencrypt": {
		"email": "jrnoble@me.com",
		"names": "meshcentral.longnoble.org",
		"production": true
	},
```
By removing those properties, the 'letsencrypt' object is now valid and will be passed into MeshCentral.

**Comment by jrnoble83** [comment]
I will give that a try, thanks for the input!

**Comment by jrnoble83** [comment]
I have tried the requested changes - the only additional change I did was change:

                            "production": false

When running node node_modules/meshcentral --debug cert   -- I am still getting a failure error:

    CERT: LE: Failed to obtain certificate: The Value "jwk" is invalid for option "format"

**Comment by si458** [comment]
@jrnoble83 please can you post your `config.json` again?
use https://jsonlint.com/ to format it correctly for us to see it OR pastebin it

**Comment by si458** [comment]
@jrnoble83 can you also post the content of your `package.json` file, need to check you have the new version installed
also what version of nodejs are you running?

**Comment by si458** [comment]
@jrnoble83 can you also please try updating your version as you are on the stable branch and could try the latest branch 1.0.85

**Comment by dinger1986** [comment]
@jrnoble83 is meshcentral.longnoble.org your address? if so its not accessible from my machine.

I would start off with https://letsdebug.net and put in your domain and see if that works, your config.json file looks to have been fine from the start as its based off the default one.

**Comment by dinger1986** [comment]
_comment

**Comment by jrnoble83** [comment]
I will get and post the requested info, in the mean time to answer your other questions.

I own longnoble.org, meshcentral.longnoble.org is one of the subdomains I'm using.  I had letsencrypt working before but it would not renew.  I removed the files in order to try to renew.  I just tried to get to it and it is working, it just has the Warning since I pulled the Cert.  The other issue you maybe having is that I have my EdgeRouter GeoIP locked down to USA only.  As far as version I had an issue, I was on latest stable but when I pushed to 1.0.85 MeshCentral would not work.  I restored my system (it's a VM on Hyper-V)  I then reverted to Stable to see if that would resolve.  Node Version is 12.22.9

**Comment by jrnoble83** [comment]
Current config.json - I also removed the relay items as I wasn't able to get it working and I stopped messing with it.  So I removed those....

{
        "$schema": "http://info.meshcentral.com/downloads/meshcentral-config-schema.json",
        "settings": {
                "MongoDb": "mongodb://127.0.0.1:27017/meshcentral",
                "cert": "meshcentral.longnoble.org",
                "_WANonly": true,
                "_LANonly": true,
                "Port": 443,
                "RedirPort": 80,
                "MpsPort": 4433,
                "_ignoreagenthashcheck": true,
                "_AllowLoginToken": true,
                "_AllowFraming": true,
                "_WebRTC": false,
                "_ClickOnce": false,
                "_UserAllowedIP": "127.0.0.1,::1,192.168.0.100",
                "Compression": true,
                "WsCompression": true,
                "AgentWsCompression": true,
                "DesktopMultiplex": true
        },
        "domains": {
                "": {
                        "title": "meshcentral.longnoble.org",
                        "_title2": "meshcentral.longnoble.org",
                        "minify": true,
                        "_newAccounts": true,
                        "_userNameIsEmail": true,
                        "mstsc": true
                }
        },
        "letsencrypt": {
                "email": "jrnoble@me.com",
                "names": "meshcentral.longnoble.org",
                "production": false
        },
        "smtp": {
                "host": "alertserver.mobilcom.com",
                "port": 465,
                "from": "meshcentral@longnoble.org",
                "user": "meshcentral@longnoble.org",
                "pass": "Jn@bl@$981983!",
                "tls": true,
                "_tlscertcheck": false,
                "_tlsstrict": true
        }
}

**Comment by dinger1986** [comment]
Theres a very high chance your edgerouter is blocking letsencrypt. letsdebug.net should tell you if that is the case.

**Comment by dinger1986** [comment]
delete the config as it has your password for your email and you probably want to change that

**Comment by jrnoble83** [comment]
I hadn't thought about that.....I'll have to remove my geoip lockdowns, thanks for that the password point out, I forgot to screen that -> I removed it

**Comment by dinger1986** [comment]
> I hadn't thought about that.....I'll have to remove my geoip lockdowns, thanks for that the password point out, I forgot to screen that -> I removed it

yeah try it, I am sure itll work as it worked before.

Theres a reasonable chance a number of people have your password in their inbox so I would still suggest changing it

**Comment by jrnoble83** [comment]
I already did change that password.  Modifying my firewall did the trick, I removed my meshcentral instance from the GeoIP restrictions and the cert renewed.  The funny thing is that I made that change more back in February or March......so I believe it renewed once since that was up, must have worked that time but blocked this time.  I'm back in business now.  Thanks for the help everyone, I was beating my head off the desk there for awhile.
